pkg/util/limiter/query_limiter.go

```diff
@@ -74,29 +74,34 @@ func QueryLimiterFromContextWithFallback(ctx context.Context) *QueryLimiter {
 }
 
 // AddSeries adds the input series and returns an error if the limit is reached.
-func (ql *QueryLimiter) AddSeries(seriesLabels labels.Labels) validation.LimitError {
-	// If the max series is unlimited just return without managing map
-	if ql.maxSeriesPerQuery == 0 {
-		return nil
-	}
+func (ql *QueryLimiter) AddSeries(seriesLabels *labels.Labels) (*labels.Labels, validation.LimitError) {
 	fingerprint := seriesLabels.Hash()
 
 	ql.uniqueSeriesMx.Lock()
 	defer ql.uniqueSeriesMx.Unlock()
 
 	uniqueSeriesBefore := len(ql.uniqueSeries)
-	ql.uniqueSeries[fingerprint] = struct{}{}
+	existing, duplicated := ql.uniqueSeries[fingerprint]
+	ql.uniqueSeries[fingerprint] = seriesLabels
 	uniqueSeriesAfter := len(ql.uniqueSeries)
 
-	if uniqueSeriesAfter > ql.maxSeriesPerQuery {
+	if ql.maxSeriesPerQuery != 0 && uniqueSeriesAfter > ql.maxSeriesPerQuery {
 		if uniqueSeriesBefore <= ql.maxSeriesPerQuery {
 			// If we've just exceeded the limit for the first time for this query, increment the failed query metric.
 			ql.queryMetrics.QueriesRejectedTotal.WithLabelValues(stats.RejectReasonMaxSeries).Inc()
 		}
 
-		return NewMaxSeriesHitLimitError(uint64(ql.maxSeriesPerQuery))
+		if duplicated {
+			return existing, NewMaxSeriesHitLimitError(uint64(ql.maxSeriesPerQuery))
+		}
+
+		return seriesLabels, NewMaxSeriesHitLimitError(uint64(ql.maxSeriesPerQuery))
```

Comment:
If we've hit the limit, then we don't need to return either set of labels:
```suggestion
		return labels.EmptyLabels(), NewMaxSeriesHitLimitError(uint64(ql.maxSeriesPerQuery))
```